README
Bigtop version 0.06
===========================

Bigtop is a central place to collect everything that has anything to
do with your database driven web application.  I call it a Web Application
Data Language.

By specifying a Bigtop file and running bigtop on it, you can generate
a working web application.  As things change (like the data model), you
can rerun bigtop to update the generated parts of the application without
fear of loosing the parts you wrote by hand.

See Bigtop::Docs::Tutorial for details.

INSTALLATION

To install this module type:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install

DEPENDENCIES

This module requires these other modules and libraries:

    Module::Build
    Config::General
    Template
    Inline
    Inline::TT

It also requires these modules during testing:

    Test::More
    Test::Warn

JOIN US

If you would like to discuss bigtop, please join the Gantry mailing list.
Go to http://www.usegantry.org and click on the Mailing List tab under the
banner for instructions.

If you want to see the most recent version of Bigtop before it hits CPAN,
you may check it out:

    svn checkout http://svn.usegantry.org/repo/bigtop/trunk/ yourdir

COPYRIGHT AND LICENCE

Copyright (c) 2005 by Phil Crow.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.



Changes
Revision history for Perl extension Bigtop.

0.33 Sat Jul 11 10:05:26 CDT 2009
    - Added ability to specify fields as being unique.
    - Added gantry backend counterpart to gantry commit r1451:
            * Skip querying of foreign tables unless a foreign key is actually
              used on the form. This gives a huge speed increase when dealing
              with foreign tables with many rows.
    - Added html_form_fieldset keyword to fields. This allows you to specify
      the fieldset to group a field into.
    - Corrected several tests so they can pass when Gantry is not installed.
    - Corrected some documentation related to kickstart syntax.
    - Refactored Kickstart.pm so you can easily subclass it to control
      what columns are placed in tables during kickstart.
    - Backends now fire in the order they are listed, except that Init
      backends are run again after the others (so they can build manifests).
    - Moved build directory validation from parser to Init:: modules.
      Default is in Init::Std.  This allows non-CPAN style builds if you
      have your own Init:: module.
    - Moved instructional messages from bigtop script to style files
      (default is in Kickstart.pm).  This means exotic build styles can
      now print their own messages.
    - Added DB2 SQL backend.
    - Fixed (I hope) an ordering bug in DBIx::Class model backend when
      using InflateColumn::DateTime.
    - Changed grammar to permit schemas in join table names.  However,
      Model backends may not yet know what to do when schemas are present.

0.32 not released to CPAN
    - Fixed long standing bug which prevented join_tables from having
      multiple data statements.

0.31 Wed Jun 11 09:37:05 CDT 2008 (Not released to CPAN)
    - Added foreign key creation for join_tables at the request of
      Stewart Heckenberg.  (Currently PostgreSQL only.)
    - Added add_columns field statement to support arbitrary entries in
      calls to it for DBIx::Class usrs (notably Kevin Esteb).
      To use it well, also add extra_components `InflateColumn::DateTime`;
      or something similar to the Model backend block in the top level config
      section.
    - Updated Gantry backend control module to work with JSON 2.x as some
      things have changed with its API. Tests were also updated to reflect
      this change.
    - Added pseudo column support
        - see docs/keyword_cookbook/field/pseudo_value for documentation
    - Update to allow labels to have single quotes in them without having
      to be escaped.

0.30 Tue Aug 14 17:15:11 CDT 2007
    - fix DBIx::Class paging routine in the generated do_main subroutines.
    - gen_root no longer creates duplicate root config vars, in fact it
      is completely ignored.  By default, all appropriate backends make
      a root whenever you don't make one yourself.
    - fixed bug with literal GantryLocations which ignored them unless
      you already had a controller level config block.

0.29 Fri Jun 22 09:23:20 CDT 2007
    - absolute paths are now used for conf file specification to aid CGIs
      you should still remove explicit conf files and use the default
      /etc/gantry.conf for production environments
    - CGI scripts are now confed according to the config CGI {} block
    - limit_by no longer causes a warning when no limit parameters is used
    - added where_terms keyword for main listing methods
    - added accessor keyword for table fields, to rename default DBIx accessor
    - corrected app.server so dbpass is not always set to dbuser's value
      thanks to Tim Allwine for pointing out the bug.
    - Stub base modules no longer use their children.
    - bigtop no longer prints instructions when no_gen tells it not to build
    - kickstart syntax no longer treats id, created, and modified as a group
      which is left out whenever you mention one of them.  Now you are
      responsible for any that you do mention and the others are still made.
    - when you have html_form_options, your stub model gets enough pod for
      its tests to pass (as long as the field is in before the stub is made)
    - bigtop with -c no longer litters on the disk if the list of things to
      build doesn't list any valid backend types (or the word all).
    - schema_base_class methods get POD to keep coverage tests happy
    - tests liter less

0.28 Fri May 11 14:55:26 CDT 2007
    - added controller_config_block.ttc to MANIFEST

0.27 Thu May 10 09:10:22 CDT 2007
    - added skipinvalid flag to json converter. ( skips the orm row )
    - added orm row to do_main data
    - added gantry conf to generation when using kickstart file
    - added permissions keyword controller hashref method type
    - added support for controller_config

        method controller_config is hashref {
            literal `mykey => 'somevalue'`;
            authed_methods 
                do_delete => ``;
            permissions `crudcrud-r--` => admin;
        }
    - added support for Gantry's new row and table level permissions new
      keywords in main listing methods:
          header_option_perms `Special Add` => create;
          row_option_perms `Exotic Edit` => update, `See Details` => retrieve;
    - Change tentmaker so it uses bigtop keywords instead of labels for
      easier movement between it an hand editing (and to consolidate book
      chapters, eventually).
    - tentmaker now allows you to enter import lists on all uses statements.
    - added controller level plugins keyword (not for base controllers,
      keep using bigtop config plugins keyword for them)
    - fixed generated default page hit tests to use top level plugins keyword
    - put a guard around t/bigtop/01_twotable.t so it won't be run if it
      can't find the Gantry templates
    - improved error message and reported line when a method block has no type
    - moved from single quotes to q!! for legends on formsets, setting an

Build.PL
use strict;

use File::Spec;
use Module::Build;

my $subclass = Module::Build->subclass(
    class   => 'Bigtop::Builder',
    code    => &_custom_code(),
);

my $build = $subclass->new(
    module_name         => 'Bigtop',
    license             => 'perl',
    create_makefile_pl  => 'passthrough',
    requires            => {
        'Parse::RecDescent' => 0,
        'Config::General' => 0,
        'Template'        => 2.0,
        'Inline'          => 0.40,
        'Inline::TT'      => 0.06,
        'Gantry'          => 3.48, # need row level permissions
        'IO::Prompt'      => 0,
        'HTTP::Server::Simple' => 0.27, # for tentmaker
        'JSON'            => 0, # because gantry controllers will want it
        'File::HomeDir'   => 0, # bigtop defaults come from home dirs
        'DBIx::Class'     => 0, # kickstarted apps assume you have this
    },
    build_requires      => {
        'Test'        => 0,
        'Test::More'  => 0,
        'Test::Warn'  => 0,
        'Test::Files' => 0.13,
    },
    script_files        => [ glob( 'scripts/*' ) ],
    tentmaker_files     => [ glob( 'tenttemplates/*' ) ],
    test_files          => [
#        't/dbixclass/01_dbixclass.t',
        't/*.t',
        't/scripthelp/*.t',
        't/bigtop/*.t',
        't/tentmaker/*.t',
        't/init/*.t',
        't/postgres/*.t',
        't/sqlite/*.t',
        't/mysql/*.t',
        't/db2/*.t',
        't/gantry/*.t',
        't/sitelook/*.t',
        't/dbixclass/*.t',
        't/cdbi/*.t',
        't/cleanup/*.t',
	],
);

eval { require Gantry; };
my $try_for_tentmaker = ( $@ ) ? 0 : 1;

if ( $try_for_tentmaker ) {

    my $default_template_path = $ENV{'TENTMAKER_TEMPLATE_PATH'} || '/usr/local/share/TentMaker';

    eval {
        require Bigtop::TentMakerPath;
        $default_template_path = Bigtop::TentMakerPath->get_template_path;
    };

    my $install_tentmaker = 'y';
    if (! $ENV{'TENTMAKER_TEMPLATE_PATH'}) {
        print "\n\nBigtop has an editor called tentmaker.  It allows\n";
        print "you to edit bigtop files with a DOM compatible browser\n";
        print "(like Firefox).  tentmaker requires some templates.\n\n";

        my $install_tentmaker = $build->y_n(
            "Do you want to install the tentmaker templates?",
            'y'
        );
    }
    
    $build->notes( install_tentmaker => $install_tentmaker );

    if ( $install_tentmaker ) {
        my $tentmaker_template_path = $default_template_path;
        if (! $ENV{'TENTMAKER_TEMPLATE_PATH'}) {
            print "\n\nTentMaker needs to store some templates on your system.\n";
            print "Please choose a location for them.\n\n";

            $tentmaker_template_path = $build->prompt(
                "Path for TentMaker templates",
                $default_template_path,
            );
        }
        
        $build->notes( tentmaker_template_path => $tentmaker_template_path );

        # Write a module which tentmaker can use to find its templates.
        my $tent_path_module = File::Spec->catfile(
                                    qw( lib Bigtop TentMakerPath.pm )
                                );

        my $template_dir;

        if ( open my $TENT_PATH, '>', $tent_path_module ) {
            print $TENT_PATH <<"EO_TENT_PATH";
package Bigtop::TentMakerPath;
use strict;

sub get_template_path {
    return '$tentmaker_template_path';
}

1;

=head1 NAME

Bigtop::TentMakerPath - keeps track of where the tentmaker templates live

=head1 SYNOPSIS

In tentmaker:

    use Bigtop::TentMakerPath;

    my \$tent_path = Bigtop::TentMakerPath->get_template_path();

=head1 DESCRIPTION

Duing initial perl Build.PL, the user is asked to supply a path for
tentmaker's templates.  If they do that, this module is written and
later installed, to keep track of where the user wanted the templates.
Then, tentmaker can call get_template_path to find out where they are.

=head1 METHODS

=over 4

=item get_template_path

Returns the path, specified by the installing user, to the tentmaker templates.

=back

=head1 AUTHOR

Phil Crow, E<lt>crow.phil\@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2006, Phil Crow

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.

=cut
EO_TENT_PATH
            close $TENT_PATH;
        }
    }
}
else {
    print "\n\n  Bigtop has an editor called tentmaker.  You need to\n";
    print "  install Gantry, AND re-run Build.PL, in order to use it.\n\n\n";
    $build->notes( install_tentmaker => 0 );
}

$build->create_build_script;

sub _custom_code {
    return q{
        sub ACTION_install {
            my $self = shift;

            $self->SUPER::ACTION_install();

            return unless ( $self->notes( 'install_tentmaker' ) );

            my $props = $self->{properties};
            my $files = $props->{tentmaker_files};

            my $path  = $self->notes( 'tentmaker_template_path' );
            foreach my $file ( @{ $files } ) {
                $self->copy_if_modified( $file, $path );
            }
            print "installed tentmaker templates to $path\n";
        }
        sub ACTION_dist {
            my $self = shift;

            if ( -x 'skip/gendocs' ) {
                `skip/gendocs`;
            }

            $self->SUPER::ACTION_dist();
        }
    };
}
